src/Toolchain.ml

```diff
@@ -0,0 +1,500 @@
+                   Some
+                     (PackageManager.esy
+                        Fpath.(projectRoot / ".vscode" / "esy"))
+                   |> Js.Promise.resolve
+               else
+                 None |> Js.Promise.resolve)
+    | file -> (
+      let manifestFile =
+        (let open Fpath in
+        projectRoot / file)
+        |> Fpath.show
+      in
+      match Path.extname file with
+      | ".json" ->
+        Fs.readFile manifestFile
+        |> Js.Promise.then_ (fun manifest ->
+               match Json.parse manifest with
+               | Some json ->
+                 if
+                   Utils.propertyExists json "dependencies"
+                   || Utils.propertyExists json "devDependencies"
+                 then
+                   Some (PackageManager.esy projectRoot) |> Js.Promise.resolve
+                 else
+                   None |> Js.Promise.resolve
+               | None ->
+                 Js.log3 "Invalid JSON file found. Ignoring..." manifest
+                   manifestFile;
+                 None |> Js.Promise.resolve)
+      | ".opam" ->
+        Fs.readFile manifestFile
+        |> Js.Promise.then_ (function
+             | "" -> None |> Js.Promise.resolve
+             | _ -> Some (PackageManager.opam projectRoot) |> Js.Promise.resolve)
+      | _ -> None |> Js.Promise.resolve )
+
+  let lookup projectRoot =
+    Fs.readDir (Fpath.toString projectRoot)
+    |> Js.Promise.then_ (function
+         | Error msg -> Js.Promise.resolve (Error msg)
+         | Ok files ->
+           files
+           |> Js.Array.map (parse projectRoot)
+           |> Js.Promise.all
+           |> Js.Promise.then_ (fun l ->
+                  Ok
+                    ( Js.Array.reduce
+                        (fun acc x ->
+                          Js.Array.concat acc
+                            ( match x with
+                            | Some x -> Array.of_list [ x ]
+                            | None -> [||] ))
+                        [||] l
+                    |> Array.to_list |> PackageManagerSet.of_list )
+                  |> Js.Promise.resolve))
+end
+
+type t =
+  { spec : PackageManager.spec
+  ; projectRoot : Fpath.t
+  }
+
+type resources =
+  { spec : PackageManager.spec
+  ; projectRoot : Fpath.t
+  }
+
+let init ~env ~folder =
+  let projectRoot = Fpath.ofString folder in
+  Js.Promise.all2
+    ( PackageManager.available ~root:projectRoot ~env
+    , Manifest.lookup projectRoot
+      |> okThen (fun pms ->
+             if pms = PackageManagerSet.empty then
+               Error "TODO: global toolchain"
+             else
+               Ok pms) )
+  |> Js.Promise.then_
+       (fun (availablePackageManagers, alreadyUsedPackageManagers) ->
+         let availablePackageManagers =
+           match availablePackageManagers with
+           | Ok x -> x |> PackageManagerSet.of_list
+           | Error msg ->
+             Js.log2 "Error during availablePackageManagers()" msg;
```

Comment:
These messages are outputted to the developer console, right? Let's use appropriate levels then? `Js.Console.error` here and below, for example.